subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	const struct bt_data *current_data = &stream->ad[stream->ad_index].data[stream->data_index];
+
+	const bool is_first_data = (stream->ad_index == 0) && (stream->data_index == 0);
+	const bool is_last_data = (stream->ad_index == stream->ad_len - 1) &&
+				  (stream->data_index == stream->ad[stream->ad_index].len - 1);
+
+	if (stream->current_data_offset == 0) {
+		/* Start reading out a new bt_data structure */
+
+		size_t max_data_to_be_read = current_data->data_len + 2;
+
+		buf[0] = current_data->data_len + 1;
+		buf[1] = current_data->type;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
```

Comment:
Given the multiple uses of `- 2` or `+ 2` (which of course is size of the length and type field`, not only in this PR, but in general in the host, we should really consider creating a macro for that value.